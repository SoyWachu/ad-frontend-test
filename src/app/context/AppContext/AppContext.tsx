"use client";
import { createContext, useContext } from "react";
import type { AppState, AppAction } from "@/app/reducers/appReducer";

export const AppContext = createContext<{
  state: AppState;
  dispatch: React.Dispatch<AppAction>;
  fetchGames: (
    page?: number,
    genre?: string,
    append?: boolean
  ) => Promise<void>;
} | null>(null);

export function useApp() {
  const context = useContext(AppContext);
  if (!context) throw new Error("useApp debe usarse dentro de AppProvider");
  return context;
}
