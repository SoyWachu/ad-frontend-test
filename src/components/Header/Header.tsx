"use client";
import Link from "next/link";
import { useApp } from "@/app/context/AppContext/AppContext";
import Image from "next/image";

export default function Header() {
  const { state } = useApp();
  return (
    <header className="flex justify-between items-center px-5 sm:px-5 lg:px-32 py-4 bg-header text-white">
      <Link href="/" className="text-2xl text-[#585660] font-bold">
        GamerShop
      </Link>
      <Link href="/cart" className="relative">
        <Image
          alt="cart"
          src="/icons/cart.png"
          className="w-6 h-6 cursor-pointer"
          width={24}
          height={24}
        />
        {state.cart.length > 0 && (
          <span className="absolute -top-2 -right-2 bg-red-500 text-white text-xs w-5 h-5 flex items-center justify-center rounded-full">
            {state.cart.length}
          </span>
        )}
      </Link>
    </header>
  );
}
