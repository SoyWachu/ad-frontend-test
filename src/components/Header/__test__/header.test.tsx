import React from "react";
import { render, screen } from "@testing-library/react";
import Header from "@/components/Header/Header";
import { useApp } from "@/app/context/AppContext/AppContext";

jest.mock("@/app/context/AppContext/AppContext", () => ({
  __esModule: true,
  useApp: jest.fn(),
}));

jest.mock("next/link", () => ({
  __esModule: true,
  default: ({ children, href }: any) => <a href={href}>{children}</a>,
}));

describe("Header", () => {
  const cartMock = [{ id: "1", title: "Game 1", quantity: 2 }];

  beforeEach(() => {
    jest.clearAllMocks();
    (useApp as jest.Mock).mockReturnValue({
      state: { cart: cartMock },
    });
  });

  it("renders logo and home link", () => {
    render(<Header />);

    const logo = screen.getByText("GamerShop");
    expect(logo).toBeInTheDocument();

    const homeLink = screen.getByRole("link", { name: /GamerShop/i });
    expect(homeLink).toHaveAttribute("href", "/");
  });

  it("renders cart icon and badge if items exist", () => {
    render(<Header />);
    const badge = screen.getByText("1");
    expect(badge).toBeInTheDocument();
  });

  it("renders cart icon without badge if cart is empty", () => {
    (useApp as jest.Mock).mockReturnValue({
      state: { cart: [] },
    });

    render(<Header />);
    const badge = screen.queryByText("1");
    expect(badge).not.toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });
});
