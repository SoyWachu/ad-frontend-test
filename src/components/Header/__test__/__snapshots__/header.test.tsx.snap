// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Header matches snapshot 1`] = `
<div>
  <header
    class="flex justify-between items-center px-5 sm:px-5 lg:px-32 py-4 bg-header text-white"
  >
    <a
      href="/"
    >
      GamerShop
    </a>
    <a
      href="/cart"
    >
      <img
        alt="cart"
        class="w-6 h-6 cursor-pointer"
        height="24"
        src="/icons/cart.png"
        width="24"
      />
      <span
        class="absolute -top-2 -right-2 bg-red-500 text-white text-xs w-5 h-5 flex items-center justify-center rounded-full"
      >
        1
      </span>
    </a>
  </header>
</div>
`;
